File: backend/src/main/java/com/ufc/backend/backend/services/ClassesService.java
package com.ufc.backend.backend.services;

import com.ufc.backend.backend.exceptions.ObjectNotFoundException;
import com.ufc.backend.backend.model.Classes;
import com.ufc.backend.backend.model.feedback.Feedback;
import com.ufc.backend.backend.repositories.ClassRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class ClassesService {
    @Autowired
    private ClassRepository repository;

    public Classes findById(String id) {
        return repository.findById(id).orElseThrow(() -> new ObjectNotFoundException(id));
    }

    public void save(Classes classes){
        repository.save(classes);
    }

    @Cacheable("classes")
    public List<Classes> findAll() {
        return repository.findAll();
    }

    public List<Classes> findAllClassesThatHasTheGivenPreRequisite(String preRequisiteId) {
        return findAll().stream().filter(obj -> obj.getPreRequisite() != null && obj.getPreRequisite().getId().equals(preRequisiteId)).collect(Collectors.toList());
    }
    public List<Feedback> findFeedbacks(String id){
        return findById(id).getFeedbacks();
    }
}

File: backend/src/main/java/com/ufc/backend/backend/services/ClassesService.kt
package com.ufc.backend.backend.services

import com.ufc.backend.backend.exceptions.ObjectNotFoundException
import com.ufc.backend.backend.model.subject.Subject
import com.ufc.backend.backend.model.feedback.Feedback
import com.ufc.backend.backend.repositories.ClassRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class ClassesService
    (
    private val repository: ClassRepository,
) {
    fun findById(id: String): Subject {
        LOGGER.info("Searching class with the given id: $id")
        return repository.findById(id).orElseThrow { ObjectNotFoundException(id) }
    }

    fun save(subject: Subject) {
        repository.save(subject)
        LOGGER.info("Saving the given class in the database: $subject")
    }

    fun findAll(): Collection<Subject> {
        LOGGER.info("Finding all subject")
        return repository.findAll()
    }

    fun findAllClassesThatHasTheGivenPreRequisite(preRequisiteId: String): Collection<Subject> {
        LOGGER.info("Finding subject that has the given pre-requisite: $preRequisiteId")
        return findAll().filter { it.preRequisite?.id == preRequisiteId }
    }

    fun findFeedbacks(id: String): Collection<Feedback> {
        LOGGER.info("Finding feedbacks of the given class: $id")
        return findById(id).feedbacks
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(ClassesService::class.java)
    }
}